src/Hpack/Render/Dependency.hs

```diff
@@ -0,0 +1,49 @@
+module Hpack.Render.Dependency (
+  renderDependencies
+, renderVersion
+) where
+
+import           Data.DList (DList)
+import qualified Data.DList as DList
+import           Data.Foldable (toList)
+import qualified Data.Map as Map
+import           Data.Semigroup (Semigroup(..))
+import           Hpack.Render.Dsl
+import           Hpack.Syntax.Dependency
+import           Hpack.Syntax.DependencyVersion
+
+data DependencyFields = DependencyFields {
+  _dependencyFieldDepends :: DList String
+, _dependencyFieldMixins :: DList String
+}
+
+instance Semigroup DependencyFields where
+  DependencyFields depsA mixinsA <> DependencyFields depsB mixinsB =
+    DependencyFields (depsA <> depsB) (mixinsA <> mixinsB)
+
+instance Monoid DependencyFields where
+  mempty = DependencyFields mempty mempty
+  mappend = (<>)
+
+renderDependencies :: String -> Dependencies -> [Element]
+renderDependencies name deps =
+  let DependencyFields renderedDeps mixins =
+        Map.foldMapWithKey renderDependency $ unDependencies deps
+  in concat
+     [ [ Field name $ CommaSeparatedList $ toList renderedDeps ]
+     , [ Field "mixins" $ CommaSeparatedList $ toList mixins
+       | not (null (toList mixins))
+       ]
+     ]
+
+renderDependency :: String -> DependencyInfo -> DependencyFields
+renderDependency name (DependencyInfo version mixins) =
+  DependencyFields
+    (DList.singleton $ name ++ renderVersion version)
+    (DList.fromList [ name ++ " " ++ mixin | mixin <- mixins ])
```

Comment:
This code is sophisticated and clever.

But if we remove the indirection over the type class it may be easier to understand for the casual reader.  It's also much shorter:
```haskell
renderDependencies :: String -> Dependencies -> [Element]
renderDependencies name deps = [
    Field name (CommaSeparatedList renderedDeps)
  , Field "mixins" (CommaSeparatedList $ concat mixins)
  ]
  where
    (renderedDeps, mixins) = unzip . map renderDependency . Map.toList $ unDependencies deps

renderDependency :: (String, DependencyInfo) -> (String, [String])
renderDependency (name, DependencyInfo version mixins) = (
      name ++ renderVersion version
    , [ name ++ " " ++ mixin | mixin <- mixins ]
    )
```